inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/kafka/src/main/java/org/apache/inlong/sort/kafka/table/ReducingUpsertWriter.java

```diff
@@ -0,0 +1,198 @@
+        this.batchIntervalMs = bufferFlushMode.getBatchIntervalMs();
+        registerFlush();
+        List<LogicalType> fields = physicalDataType.getLogicalType().getChildren();
+        final RowData.FieldGetter[] keyFieldGetters =
+                Arrays.stream(keyProjection)
+                        .mapToObj(
+                                targetField -> RowData.createFieldGetter(
+                                        fields.get(targetField), targetField))
+                        .toArray(RowData.FieldGetter[]::new);
+        this.keyExtractor = rowData -> createProjectedRow(rowData, RowKind.INSERT, keyFieldGetters);
+        this.valueCopyFunction = valueCopyFunction;
+        if (metricOption != null) {
+            this.sourceMetricData = new SourceMetricData(metricOption);
+        }
+    }
+
+    @Override
+    public void write(RowData element, Context context) throws IOException, InterruptedException {
+        wrappedContext.setContext(context);
+        addToBuffer(element, context.timestamp());
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException, InterruptedException {
+        flush();
+    }
+
+    @Override
+    public List<WriterState> snapshotState(long checkpointId) throws IOException {
+        return wrappedWriter.snapshotState(checkpointId);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!closed) {
+            closed = true;
+            wrappedWriter.close();
+        }
+    }
+
+    private void addToBuffer(RowData row, Long timestamp) throws IOException, InterruptedException {
+        RowData key = keyExtractor.apply(row);
+        RowData value = valueCopyFunction.apply(row);
+        reduceBuffer.put(key, new Tuple2<>(changeFlag(value), timestamp));
+
+        if (reduceBuffer.size() >= batchMaxRowNums) {
+            flush();
+        }
+    }
+
+    private void registerFlush() {
+        if (closed) {
+            return;
+        }
+        timeService.registerTimer(
+                lastFlush + batchIntervalMs,
+                (t) -> {
+                    if (t >= lastFlush + batchIntervalMs) {
+                        flush();
+                    }
+                    registerFlush();
+                });
+    }
+
+    private RowData changeFlag(RowData value) {
+        switch (value.getRowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                value.setRowKind(UPDATE_AFTER);
+                break;
+            case UPDATE_BEFORE:
+            case DELETE:
+                value.setRowKind(DELETE);
+        }
+        return value;
+    }
+
+    private void flush() throws IOException, InterruptedException {
+        for (Tuple2<RowData, Long> value : reduceBuffer.values()) {
+            wrappedContext.setTimestamp(value.f1);
+            wrappedWriter.write(value.f0, wrappedContext);
+            sourceMetricData.outputMetricsWithEstimate(value.f0);
```

Comment:
NPE may happen if `sourceMetricData` is  null

Reply:
fixed it. thx~